File: src/test/java/com/machfour/macros/core/CsvTest.java
package com.machfour.macros.core;

import com.machfour.macros.linux.Config;
import com.machfour.macros.objects.Food;
import com.machfour.macros.objects.NutritionData;
import com.machfour.macros.objects.Serving;
import com.machfour.macros.storage.CsvStorage;
import com.machfour.macros.linux.LinuxDatabase;
import com.machfour.macros.storage.MacrosDatabase;
import com.machfour.macros.util.Pair;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.*;
import java.sql.SQLException;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class CsvTest {
    static LinuxDatabase db;
    static final String TEST_DB_LOCATION = "/home/max/devel/macros-java/test.sqlite";
    static final String REAL_DB_LOCATION = "/home/max/devel/macros-java/macros.sqlite";
    static final String TEST_WRITE_DIR = "/home/max/devel/macros-java/test";

    @BeforeAll
    static void initDb() {
        db = LinuxDatabase.getInstance(TEST_DB_LOCATION);
        try {
            db.deleteIfExists(TEST_DB_LOCATION);
            db.initDb();
        } catch (IOException e1) {
            e1.printStackTrace();
            fail("Database initialisation threw IO exception");
        } catch (SQLException e2) {
            e2.printStackTrace();
            fail("Database initialisation threw SQL exception");
        }
    }

    @Test
    void testCsvReadFoods() {
        Map<String, Food> csvFoods;
        try (Reader foodCsv = new FileReader(Config.FOOD_CSV_FILENAME) ){
            csvFoods = CsvStorage.buildFoodObjectTree(foodCsv);
            assertNotEquals(0, csvFoods.size(), "CSV read in zero foods!");
            //System.out.println(csvObjects.first.get(0));
        } catch (IOException e) {
            e.printStackTrace();
            fail("IOException was thrown");
        }
    }
    @Test
    void testCsvReadServings() {
        List<Serving> csvServings;
        try (Reader servingCsv = new FileReader(Config.SERVING_CSV_FILENAME) ){
            csvServings = CsvStorage.buildServings(servingCsv);
            assertNotEquals(0, csvServings.size(), "CSV read in zero servings!");
            System.out.println(csvServings.get(0));
        } catch (IOException e) {
            e.printStackTrace();
            fail("IOException was thrown");
        }
    }
    @Test
    void testCsvSaveFoods() {
        Pair<List<Food>, List<NutritionData>> csvObjects;
        try (Reader foodCsv = new FileReader(Config.FOOD_CSV_FILENAME) ){
            CsvStorage.importFoodData(foodCsv, db, true);
        } catch (IOException e) {
            e.printStackTrace();
            fail("IOException was thrown");
        } catch (SQLException e2) {
            e2.printStackTrace();
            fail("Database save threw SQL exception");
        }
    }
    @Test
    void testCsvSaveServings() {
        List<Serving> csvServings;
        try (Reader servingCsv = new FileReader(Config.SERVING_CSV_FILENAME)) {
            CsvStorage.importServings(servingCsv, db, true);
        } catch (IOException e) {
            e.printStackTrace();
            fail("IOException was thrown");
        } catch (SQLException e2) {
            e2.printStackTrace();
            fail("Database save threw SQL exception");
        }
    }
    @Test
    void testCsvWriteFoods() {
        MacrosDatabase db = LinuxDatabase.getInstance(REAL_DB_LOCATION);
        try (Writer csvOut = new FileWriter(TEST_WRITE_DIR + "/all-food.csv")){
            Map<Long, Food> foods = db.getAllRawObjects(Food.table());
            CsvStorage.writeObjectsToCsv(Food.table(), csvOut, foods.values());
        } catch (IOException e) {
            e.printStackTrace();
            fail("IOException was thrown");
        } catch (SQLException e2) {
            e2.printStackTrace();
            fail("Database save threw SQL exception");
        }
    }
    @Test
    void testCsvWriteServings() {
        MacrosDatabase db = LinuxDatabase.getInstance(REAL_DB_LOCATION);
        try (Writer csvOut = new FileWriter(TEST_WRITE_DIR + "/all-serving.csv")) {
            Map<Long, Serving> servings = db.getAllRawObjects(Serving.table());
            CsvStorage.writeObjectsToCsv(Serving.table(), csvOut, servings.values());
        } catch (IOException e) {
            e.printStackTrace();
            fail("IOException was thrown");
        } catch (SQLException e2) {
            e2.printStackTrace();
            fail("Database save threw SQL exception");
        }
    }

    @BeforeEach
    void setUp() {
    }

    @AfterEach
    void tearDown() {
    }
}

File: src/test/java/storage/CsvTest.kt
package storage

import com.machfour.macros.core.MacrosConfig
import com.machfour.macros.core.datatype.TypeCastException
import com.machfour.macros.linux.LinuxConfig
import com.machfour.macros.linux.LinuxDatabase
import com.machfour.macros.linux.LinuxDatabase.Companion.deleteIfExists
import com.machfour.macros.linux.LinuxDatabase.Companion.getInstance
import com.machfour.macros.objects.*
import com.machfour.macros.queries.FoodPortionQueries
import com.machfour.macros.storage.CsvBackup.writeObjectsToCsv
import com.machfour.macros.storage.CsvImport.buildFoodObjectTree
import com.machfour.macros.storage.CsvImport.buildServings
import com.machfour.macros.storage.CsvImport.importFoodData
import com.machfour.macros.storage.CsvImport.importRecipes
import com.machfour.macros.storage.CsvImport.importServings
import com.machfour.macros.validation.SchemaViolation
import org.junit.jupiter.api.*
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.sql.SQLException

class CsvTest {
    companion object {
        lateinit var db: LinuxDatabase

        // TODO make test config
        val config: MacrosConfig = LinuxConfig()
        const val TEST_DB_LOCATION = "/home/max/devel/macros/test.sqlite"
        const val REAL_DB_LOCATION = "/home/max/devel/macros/macros.sqlite"
        const val TEST_WRITE_DIR = "/home/max/devel/macros/test"

        @JvmStatic
        @BeforeAll
        fun initDb() {
            try {
                db = getInstance(TEST_DB_LOCATION)
                deleteIfExists(TEST_DB_LOCATION)
                db.initDb()
            } catch (e1: IOException) {
                e1.printStackTrace()
                Assertions.fail<Any>("Database initialisation threw IO exception")
            } catch (e2: SQLException) {
                e2.printStackTrace()
                Assertions.fail<Any>("Database initialisation threw SQL exception")
            }
        }

    }

    @BeforeEach
    fun clearDb() {
        FoodPortionQueries.deleteAllIngredients(db)
        db.clearTable(Serving.table)
        db.clearTable(NutrientValue.table)
        db.clearTable(Food.table)
    }

    @Test
    fun testCsvReadFoods() {
        var csvFoods: Map<String, Food>
        try {
            FileReader(config.foodCsvPath).use {
                csvFoods = buildFoodObjectTree(it)
                Assertions.assertNotEquals(0, csvFoods.size, "CSV read in zero foods!")
            }
        } catch (e: IOException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: TypeCastException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        }
    }

    @Test
    fun testCsvReadServings() {
        var csvServings: List<Serving>
        try {
            FileReader(config.servingCsvPath).use {
                csvServings = buildServings(it)
                Assertions.assertNotEquals(0, csvServings.size, "CSV read in zero servings!")
                println(csvServings[0])
            }
        } catch (e: IOException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: TypeCastException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        }
    }

    @Test
    fun testCsvSaveFoods() {
        try {
            FileReader(config.foodCsvPath).use { importFoodData(db, it, true) }
        } catch (e: SQLException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: IOException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: TypeCastException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        }
    }

    @Test
    fun testCsvSaveServings() {
        try {
            // save foods first
            FileReader(config.foodCsvPath).use { importFoodData(db, it, true) }
            FileReader(config.servingCsvPath).use { importServings(db, it, true) }
        } catch (e: SQLException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: IOException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: TypeCastException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: SchemaViolation) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        }
    }

    @Test
    fun testCsvWriteFoods() {
        try {
            val foods = db.getAllRawObjects(Food.table)
            FileWriter("$TEST_WRITE_DIR/all-food.csv").use { writeObjectsToCsv(Food.table, it, foods.values) }
        } catch (e: IOException) {
            e.printStackTrace()
            Assertions.fail<Any>("IOException was thrown")
        } catch (e2: SQLException) {
            e2.printStackTrace()
            Assertions.fail<Any>("Database save threw SQL exception")
        }
    }

    @Test
    fun testCsvWriteServings() {
        try {
            val servings = db.getAllRawObjects(Serving.table)
            FileWriter("$TEST_WRITE_DIR/all-serving.csv").use { writeObjectsToCsv(Serving.table, it, servings.values) }
        } catch (e: SQLException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: IOException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        }
    }

    // put this test after other imports
    @Test
    fun testCsvSaveRecipes() {
        try {
            FileReader(config.foodCsvPath).use { importFoodData(db, it, true) }
            FileReader(config.recipeCsvPath).use { recipeCsv ->
                FileReader(config.ingredientsCsvPath).use { ingredientCsv ->
                    importRecipes(db, recipeCsv, ingredientCsv)
                }
            }
        } catch (e: SQLException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: IOException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        } catch (e: TypeCastException) {
            e.printStackTrace()
            Assertions.fail<Any>("Exception was thrown")
        }
    }

    @BeforeEach
    fun setUp() {
    }

    @AfterEach
    fun tearDown() {
    }
}